File: packages/launcher_domain/lib/src/launch_resolution/launch_resolver.dart
part of '../launch_resolution.dart';

abstract final class LaunchResolver {
  static LaunchResolution resolve(
    EffectiveProfile profile,
    LaunchRequest request, [
    RuntimeObservation observation = RuntimeObservation.none,
    RuntimeBindingSnapshot? bindings,
  ]) {
    final blocks = <LaunchBlock>[];
    final index = _OwnerIndex(profile);
    blocks.addAll(index.ambiguities);
    final target = _findTarget(index, request.targetId, blocks);
    if (target == null) return LaunchResolution.blocked(blocks);
    if (bindings != null && !bindings.matches(profile)) {
      return LaunchResolution.blocked([
        LaunchBlock(
          LaunchBlockCode.planPackageSetMismatch,
          target.declaration.id,
        ),
      ]);
    }
    if (!_supportsThisInstall(
      profile.install,
      index.manifest(target.package),
    )) {
      blocks.add(
        LaunchBlock(
          LaunchBlockCode.targetPlatformUnsupported,
          target.package.id,
          target.package.version,
        ),
      );
    }
    final validObservation = observation.matches(profile)
        ? observation
        : RuntimeObservation.none;
    final gamemode = _resolveGamemode(
      index,
      target,
      validObservation,
      bindings,
      blocks,
    );
    _validatePolicyReferences(index, target, blocks);
    final world = _resolveWorld(
      index,
      profile,
      target,
      request,
      validObservation,
      bindings,
      blocks,
    );
    String transition = '';
    if (gamemode != null && world != null) {
      transition = _resolveTransition(target, gamemode, world, request, blocks);
      _checkConsent(index, target, gamemode, world, blocks);
      if (blocks.every(
            (block) => block.code == LaunchBlockCode.worldUnavailable,
          ) &&
          _allCandidatesUnavailable(
            index,
            profile,
            target,
            gamemode,
            request,
            validObservation,
          )) {
        blocks.add(
          LaunchBlock(LaunchBlockCode.noAvailableTarget, target.declaration.id),
        );
      }
    }
    if (blocks.isNotEmpty) return LaunchResolution.blocked(blocks);
    return LaunchResolution.success(
      LaunchPlan._(
        targetId: target.declaration.id,
        gamemodeId: gamemode!.declaration.id,
        worldId: world!.instanceId ?? world.declaration.id,
        worldFamilyId: world.instanceId == null ? null : world.declaration.id,
        transition: transition,
        request: request,
        packages: profile.packages,
        target: target.declaration,
        gamemode: gamemode.declaration,
        world: world.declaration,
      ),
    );
  }

  static List<LaunchBlock> revalidate(
    LaunchPlanDescriptor plan,
    Iterable<PackageIdentity> loaded,
  ) {
    final copy = loaded.toList();
    return _samePackages(plan.packages, copy) &&
            packageSetDigest(copy) == plan.digest
        ? const []
        : [LaunchBlock(LaunchBlockCode.planPackageSetMismatch, plan.targetId)];
  }

  static LaunchResolution resolveAgain(
    LaunchPlanDescriptor plan,
    EffectiveProfile loaded, [
    RuntimeObservation observation = RuntimeObservation.none,
    RuntimeBindingSnapshot? bindings,
  ]) {
    final mismatch = revalidate(plan, loaded.packages);
    if (mismatch.isNotEmpty) return LaunchResolution.blocked(mismatch);
    final resolved = resolve(
      loaded,
      plan.request,
      observation,
      bindings ?? RuntimeBindingSnapshot._missing(loaded),
    );
    final next = resolved.plan;
    if (next == null) return resolved;
    if (next.targetId != plan.targetId ||
        next.gamemodeId != plan.gamemodeId ||
        next.worldId != plan.worldId ||
        next.worldFamilyId != plan.worldFamilyId ||
        next.transition != plan.transition) {
      return LaunchResolution.blocked([
        LaunchBlock(LaunchBlockCode.planResolutionMismatch, plan.targetId),
      ]);
    }
    return resolved;
  }
}

class _TargetMatch {
  const _TargetMatch(this.declaration, this.package);
  final ModLaunchTargetDeclaration declaration;
  final ResolvedPackage package;
}

class _GamemodeMatch {
  const _GamemodeMatch(this.declaration, this.package);
  final ModGamemodeDeclaration declaration;
  final ResolvedPackage package;
}

class _WorldMatch {
  const _WorldMatch(this.declaration, this.package, [this.instanceId]);
  final ModWorldDeclaration declaration;
  final ResolvedPackage package;
  final String? instanceId;
  String get id => instanceId ?? declaration.id;
}

_TargetMatch? _findTarget(
  _OwnerIndex index,
  String id,
  List<LaunchBlock> blocks,
) {
  final owner = index.owner(id, blocks);
  if (owner?.ambiguous == true) return null;
  final package = owner?.enabled;
  if (package == null) {
    final off = owner?.disabled;
    blocks.add(
      off == null
          ? LaunchBlock(LaunchBlockCode.targetNotDeclared, id)
          : LaunchBlock(
              LaunchBlockCode.targetPackageDisabled,
              off.id,
              off.version,
            ),
    );
    return null;
  }
  final targets =
      (index.manifest(package).contributions?.launchTargets ??
              const <ModLaunchTargetDeclaration>[])
          .where((item) => _idEquals(item.id, id))
          .toList();
  if (targets.length > 1) {
    blocks.add(LaunchBlock(LaunchBlockCode.declarationIdAmbiguous, id));
    return null;
  }
  if (targets.isNotEmpty) return _TargetMatch(targets.single, package);
  blocks.add(LaunchBlock(LaunchBlockCode.targetNotDeclared, id));
  return null;
}

_GamemodeMatch? _resolveGamemode(
  _OwnerIndex index,
  _TargetMatch target,
  RuntimeObservation observation,
  RuntimeBindingSnapshot? bindings,
  List<LaunchBlock> blocks,
) {
  final id = target.declaration.gamemode;
  final owner = index.owner(id, blocks);
  if (owner?.ambiguous == true) return null;
  final package = owner?.enabled;
  if (package == null) {
    final off = owner?.disabled;
    blocks.add(
      off == null
          ? LaunchBlock(LaunchBlockCode.gamemodeNotDeclared, id)
          : LaunchBlock(
              LaunchBlockCode.gamemodePackageDisabled,
              off.id,
              off.version,
            ),
    );
    return null;
  }
  _checkReference(
    index,
    target.package,
    package,
    id,
    LaunchBlockCode.gamemodeRefNotADependency,
    blocks,
  );
  final modes =
      index.manifest(package).contributions?.gamemodes ??
      const <ModGamemodeDeclaration>[];
  final matching = modes.where((mode) => _idEquals(mode.id, id)).toList();
  if (matching.length > 1) {
    blocks.add(LaunchBlock(LaunchBlockCode.declarationIdAmbiguous, id));
    return null;
  }
  for (final mode in matching) {
    if (!_supportsThisInstall(index.profile.install, index.manifest(package))) {
      blocks.add(
        LaunchBlock(
          LaunchBlockCode.gamemodePlatformUnsupported,
          package.id,
          package.version,
        ),
      );
    }
    final fresh = bindings != null && bindings.matches(index.profile);
    final unavailable = bindings == null
        ? observation.unboundGamemodeIds.any((item) => _idEquals(item, mode.id))
        : fresh &&
              !bindings.boundGamemodeIds.any(
                (item) => _idEquals(item, mode.id),
              );
    if (unavailable) {
      blocks.add(
        LaunchBlock(LaunchBlockCode.gamemodeUnbound, mode.id, package.version),
      );
    }
    if (fresh) _bindingFailures(bindings, 'gamemode', mode.id, blocks);
    for (final record in observation.availability) {
      if (record.kind == 'gamemode' && _idEquals(record.id, mode.id)) {
        blocks.addAll(
          record.blocks.where(
            (block) =>
                !(fresh && block.code == LaunchBlockCode.gamemodeUnbound),
          ),
        );
      }
    }
    return _GamemodeMatch(mode, package);
  }
  blocks.add(LaunchBlock(LaunchBlockCode.gamemodeNotDeclared, id));
  return null;
}

void _bindingFailures(
  RuntimeBindingSnapshot? bindings,
  String kind,
  String id,
  List<LaunchBlock> blocks,
) {
  if (bindings == null) return;
  for (final record in bindings.availability) {
    if (record.kind == kind && _idEquals(record.id, id)) {
      blocks.addAll(record.blocks);
    }
  }
}

bool _idEquals(String left, String right) =>
    left.toLowerCase() == right.toLowerCase();
bool _isDiscovered(ModWorldDeclaration world) =>
    world.content?.kind == ModWorldContent.discoveredKind;

File: src/TopiaForge.ModManager.Core/LaunchResolver.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace TopiaForge.ModManager.Core
{
    /// <summary>Pure resolution over exact installed selections; declarations and observations cannot change ownership.</summary>
    public static partial class LaunchResolver
    {
        public static LaunchResolution Resolve(EffectiveProfile profile, LaunchRequest request, RuntimeObservation? observation = null,
            RuntimeBindingSnapshot? bindings = null) =>
            ResolveCore(profile, request, observation, bindings, requireBindings: bindings != null);

        public static IReadOnlyList<LaunchBlock> Revalidate(LaunchPlan plan, IEnumerable<PackageIdentity> loaded) =>
            Revalidate(plan.Descriptor, loaded);

        public static IReadOnlyList<LaunchBlock> Revalidate(LaunchPlanDescriptor plan, IEnumerable<PackageIdentity> loaded)
        {
            var identities = loaded.ToArray();
            return LaunchContractValues.SamePackages(plan.Packages, identities) && PackageSetDigest.Of(identities) == plan.Digest
                ? Array.Empty<LaunchBlock>()
                : new[] { new LaunchBlock(LaunchBlockCode.PlanPackageSetMismatch, plan.TargetId) };
        }

        public static LaunchResolution ResolveAgain(LaunchPlanDescriptor plan, EffectiveProfile loaded,
            RuntimeObservation? observation = null, RuntimeBindingSnapshot? bindings = null)
        {
            var mismatch = Revalidate(plan, loaded.Packages.Select(package => package.Identity));
            if (mismatch.Count > 0) return LaunchResolution.Blocked(mismatch);
            if (bindings != null && !BindingsMatch(bindings, loaded))
                return LaunchResolution.Blocked(new[] { new LaunchBlock(LaunchBlockCode.PlanPackageSetMismatch, plan.TargetId) });
            var resolution = ResolveCore(loaded, plan.Request, observation, bindings, requireBindings: true);
            if (!resolution.Resolved) return resolution;
            var actual = resolution.Plan!.Descriptor;
            return actual.TargetId == plan.TargetId && actual.GamemodeId == plan.GamemodeId && actual.WorldId == plan.WorldId
                && actual.WorldFamilyId == plan.WorldFamilyId && actual.Transition == plan.Transition
                ? resolution : LaunchResolution.Blocked(new[] { new LaunchBlock(LaunchBlockCode.PlanResolutionMismatch, plan.TargetId) });
        }

        private static LaunchResolution ResolveCore(EffectiveProfile profile, LaunchRequest request, RuntimeObservation? observation,
            RuntimeBindingSnapshot? bindings, bool requireBindings)
        {
            if (profile == null) throw new ArgumentNullException(nameof(profile));
            if (request == null) throw new ArgumentNullException(nameof(request));
            if (bindings != null && !BindingsMatch(bindings, profile))
                return LaunchResolution.Blocked(new[] { new LaunchBlock(LaunchBlockCode.PlanPackageSetMismatch, request.TargetId) });
            var index = new LaunchProfileIndex(profile);
            if (index.Duplicates.Count > 0) return LaunchResolution.Blocked(index.Duplicates);
            var blocks = new List<LaunchBlock>();
            var observed = observation?.ForProfile(profile) ?? RuntimeObservation.None;
            var target = FindTarget(index, request.TargetId, blocks);
            if (target == null) return LaunchResolution.Blocked(blocks);
            if (!SupportsThisInstall(profile.Install, target.Package.Snapshot))
                blocks.Add(new LaunchBlock(LaunchBlockCode.TargetPlatformUnsupported, target.Package.Id, target.Package.Version));
            var mode = FindGamemode(index, target.Declaration.Gamemode, target.Package,
                LaunchBlockCode.GamemodeRefNotADependency, blocks);
            if (mode != null)
            {
                ApplyModeAvailability(mode, observed, bindings, requireBindings, blocks);
                if (!SupportsThisInstall(profile.Install, mode.Package.Snapshot))
                    blocks.Add(new LaunchBlock(LaunchBlockCode.GamemodePlatformUnsupported, mode.Package.Id, mode.Package.Version));
            }
            var policy = target.Declaration.World;
            if (policy == null)
            {
                blocks.Add(new LaunchBlock(LaunchBlockCode.WorldNotDeclared, target.Declaration.Id));
                return LaunchResolution.Blocked(blocks);
            }

            foreach (var reference in new[] { policy.Default }.Concat(policy.Allow).Distinct(StringComparer.OrdinalIgnoreCase))
            {
                var owner = index.OwnerOf(reference);
                if (owner?.Package != null && !owner.Disabled && !owner.Ambiguous)
                    CheckReference(target.Package, owner.Package, reference, LaunchBlockCode.WorldRefNotADependency, blocks);
                FindWorld(index, reference, observed, allowInstance: false, blocks);
            }
            var requested = request.WorldOverride ?? policy.Default;
            if (request.WorldOverride != null && (policy.Policy == ModWorldPolicy.FixedPolicy || policy.AllowPlayerOverride != true))
                blocks.Add(new LaunchBlock(LaunchBlockCode.WorldNotAdmittedByPolicy, requested));
            var world = FindWorld(index, requested, observed, allowInstance: request.WorldOverride != null, blocks);
            var transition = string.Empty;
            if (world != null)
            {
                foreach (var consent in world.Declaration.OpenTo ?? Enumerable.Empty<string>())
                    FindGamemode(index, consent, world.Package, LaunchBlockCode.WorldConsentRefNotADependency, blocks);
                if (!Admitted(policy, world, target.Declaration.Gamemode))
                    blocks.Add(new LaunchBlock(policy.Policy == ModWorldPolicy.OpenPolicy
                        ? LaunchBlockCode.WorldConsentMissing : LaunchBlockCode.WorldNotAdmittedByPolicy, world.WorldId));
                ApplyWorldAvailability(world, observed, bindings, requireBindings, blocks);
                if (!SupportsThisInstall(profile.Install, world.Package.Snapshot))
                    blocks.Add(new LaunchBlock(LaunchBlockCode.WorldPlatformUnsupported, world.Package.Id, world.Package.Version));
                if (mode != null)
                {
                    transition = ResolveTransition(target.Declaration, mode.Declaration, world, request, blocks);
                    if (blocks.All(block => block.Code == LaunchBlockCode.WorldUnavailable)
                        && observed.ExplicitlyUnavailable(world.WorldId)
                        && EveryCandidateExplicitlyUnavailable(index, profile.Install, target, mode, world, request, observed))
                        blocks.Add(new LaunchBlock(LaunchBlockCode.NoAvailableTarget, target.Declaration.Id));
                }
            }
            if (blocks.Count > 0 || mode == null || world == null) return LaunchResolution.Blocked(blocks);
            var descriptor = new LaunchPlanDescriptor(target.Declaration.Id, mode.Declaration.Id, world.WorldId, transition,
                request, profile.Packages.Select(package => package.Identity), world.FamilyId);
            return LaunchResolution.Success(new LaunchPlan(descriptor, target.Declaration, mode.Declaration, world.Declaration));
        }

        private static TargetMatch? FindTarget(LaunchProfileIndex index, string id, List<LaunchBlock> blocks)
        {
            var owner = index.ResolveOwner(id, LaunchBlockCode.TargetNotDeclared, LaunchBlockCode.TargetPackageDisabled, blocks);
            if (owner == null) return null;
            var matches = owner.Snapshot.Contributions?.LaunchTargets.Where(item => Same(item.Id, id)).ToArray()
                ?? Array.Empty<ModLaunchTargetDeclaration>();
            if (matches.Length != 1)
            {
                blocks.Add(new LaunchBlock(matches.Length > 1 ? LaunchBlockCode.DeclarationIdAmbiguous : LaunchBlockCode.TargetNotDeclared, id));
                return null;
            }
            return new TargetMatch(matches[0], owner);
        }

        private static GamemodeMatch? FindGamemode(LaunchProfileIndex index, string id, ResolvedPackage referrer,
            LaunchBlockCode dependencyCode, List<LaunchBlock> blocks)
        {
            var owner = index.ResolveOwner(id, LaunchBlockCode.GamemodeNotDeclared, LaunchBlockCode.GamemodePackageDisabled, blocks);
            if (owner == null) return null;
            CheckReference(referrer, owner, id, dependencyCode, blocks);
            var matches = owner.Snapshot.Contributions?.Gamemodes.Where(item => Same(item.Id, id)).ToArray()
                ?? Array.Empty<ModGamemodeDeclaration>();
            if (matches.Length != 1)
            {
                blocks.Add(new LaunchBlock(matches.Length > 1 ? LaunchBlockCode.DeclarationIdAmbiguous : LaunchBlockCode.GamemodeNotDeclared, id));
                return null;
            }
            return new GamemodeMatch(matches[0], owner);
        }

        private static bool Same(string? a, string? b) => LaunchProfileIndex.Same(a, b);
        private sealed class TargetMatch
        {
            internal TargetMatch(ModLaunchTargetDeclaration declaration, ResolvedPackage package) { Declaration = declaration; Package = package; }
            internal ModLaunchTargetDeclaration Declaration { get; }
            internal ResolvedPackage Package { get; }
        }
        private sealed class GamemodeMatch
        {
            internal GamemodeMatch(ModGamemodeDeclaration declaration, ResolvedPackage package) { Declaration = declaration; Package = package; }
            internal ModGamemodeDeclaration Declaration { get; }
            internal ResolvedPackage Package { get; }
        }
        private sealed class WorldMatch
        {
            internal WorldMatch(ModWorldDeclaration declaration, ResolvedPackage package, string worldId, string? familyId = null)
            { Declaration = declaration; Package = package; WorldId = worldId; FamilyId = familyId; }
            internal ModWorldDeclaration Declaration { get; }
            internal ResolvedPackage Package { get; }
            internal string WorldId { get; }
            internal string? FamilyId { get; }
        }
    }
}
